### DISCUSSION TITLE
Open ContextMenu programatically

### DISCUSSION THREAD

**Discussion by JaggerJo** [AvaloniaUI/Avalonia#15118] [Q&A]
Context:

I have a control that renders elements using an `ICustomRenderOperation`. When the user right clicks on one of the elements I want to display a context menu at that location. If a user right clicks above no element nothing is supposed to happen.

Problem:

I can't figure out how to manually open and position a context menu. 

What I've tried:

I tried embedding a `ContextMenu` in a `Popup` but this leads to strange behaviour. (Popup stays open when a MenuItem is clicked, normal ContextMenu would close)

**Comment by timunie** [comment]
Use ContextFlyout instead

**Reply by JaggerJo** [reply]
Hmm, I don't see how that would help. Could you elaborate on why this makes sense?

**Reply by timunie** [reply]
ContextFlyout can have a MenuFlyout. Its a newer API and should be used over ContextMenu for new Apps.

**Reply by JaggerJo** [reply]
Aha, good to know. 

I've tried it but it's not a great fit for my problem. The issue is that I don't handle the right click myself. I just display the context menu **IF** the user right clicked over an element. If the user right clicks empty space nothing happens.

This is what I currently have. The only remaining issue with this is that the popup stays open even if the user clicks a menu item, but this should be easy to fix by implementing something [along the lines of this](https://github.com/AvaloniaUI/Avalonia/blob/bb6005b3242731cdbacdfe94463b9d99d915953c/src/Avalonia.Controls/ContextMenu.cs#L274-L285).

Example:
```xaml
    <Panel>
        <mainEditor:MainEditorRenderView
            ContextMenuCommand="{Binding ContextMenuCommand}"/>

        <Popup
            IsOpen="{Binding ContextMenu, Converter={x:Static ObjectConverters.IsNotNull}}"
            IsLightDismissEnabled="True"
            Placement="TopEdgeAlignedLeft"
            VerticalOffset="{Binding ContextMenu.ScreenPosition.Y}"
            HorizontalOffset="{Binding ContextMenu.ScreenPosition.X}">

            <ContentControl
                Content="{Binding ContextMenu}">
                <ContentControl.DataTemplates>
                    <DataTemplate x:DataType="mainEditor:NodeContextMenuViewModel">
                        <mainEditor:NodeContextMenu/>
                    </DataTemplate>
                </ContentControl.DataTemplates>
            </ContentControl>
        </Popup>

    </Panel>
```

**Reply by timunie** [reply]
> If the user right clicks empty space nothing happens.

that's expected. You need a solid background for hittesting, `Transparent` is okay but not `null`. Maybe you can alos override hit-testing to configure that.

**Reply by JaggerJo** [reply]
>  If the user right clicks empty space nothing happens.

Sorry seems like I formulated that not too well. That's what I expect to happen / desired behaviour.

`MainEditorRenderView` does all the hit testing, .. internally and invokes `ContextMenuCommand` with information about where to place the context menu relative to itself and what kind of context menu to show. 

What I now want to do:

Show a context menu that is not triggered by right clicking. I want to control in code where to show it and when. I currently do this by wrapping the `ContextMenu` in a `Popup`. 

The only issue with that implementation is that the `Popup` remains open when menu items are clicked in the context menu.

That's why I'm looking at [replicating what context menus do](https://github.com/AvaloniaUI/Avalonia/blob/bb6005b3242731cdbacdfe94463b9d99d915953c/src/Avalonia.Controls/ContextMenu.cs#L306-L356).

**Reply by JaggerJo** [reply]
Seems like closed/closing is never invoked if the context menu is not shown via `Open(..)`

```csharp
    protected override void OnAttachedToLogicalTree(LogicalTreeAttachmentEventArgs e)
    {
        base.OnAttachedToLogicalTree(e);

        _contextMenu.Closed += (sender, args) => Console.WriteLine("Close context menu");
        _contextMenu.Closing += (sender, args) => Console.WriteLine("Closing context menu");
    }
```

**Reply by timunie** [reply]
I still would use a MenuFlyout:

```cs
protected override void OnLoaded(RoutedEventArgs e)
{
    base.OnLoaded(e);

    var flyout = new MenuFlyout()
    {
        Items =
        {
            new MenuItem() {Header = "Hello"},
            new MenuItem() {Header = "Wolrd"},
        },
        Placement = PlacementMode.Center,

    };
    
    flyout.Closing += FlyoutOnClosing;
    
    flyout.ShowAt(this);
}

private void FlyoutOnClosing(object? sender, CancelEventArgs e)
{
    Trace.WriteLine("Closing Me");
}
```

**Comment by JaggerJo** [comment]
Seems like I have to mimic what's done here:

https://github.com/AvaloniaUI/Avalonia/blob/bb6005b3242731cdbacdfe94463b9d99d915953c/src/Avalonia.Controls/ContextMenu.cs#L306-L356

From looking at the ContextMenu implementation it seems like the popup is closed when it's child ContextMenu is closed.

https://github.com/AvaloniaUI/Avalonia/blob/bb6005b3242731cdbacdfe94463b9d99d915953c/src/Avalonia.Controls/ContextMenu.cs#L274-L285

**Comment by slavslavin** [comment]
I have the same issue. Have you found a solution yet?

**Reply by timunie** [reply]
Have you tried a MenuFlyot yet?
